### ISSUE TITLE
HTTP client options

### ISSUE THREAD

**Issue by danschultzer** [honeybadger-io/honeybadger-elixir#467]
I've experienced issues with honeybadger not working in a IPv6 only network. This is due to a long running issue in hackney of how IP addresses are resolved: https://github.com/benoitc/hackney/issues/206 (in [particular this logic](https://github.com/benoitc/hackney/blob/6e79b2bb11a77389d3ba9ff3a0828a45796fe7a8/src/hackney_util.erl#L81-L91)).

Looking at how honeybadger deals with HTTP it would be good to:

1. Allow options to be passed to hackney (~~this won't solve my particular issue though~~ edit: should work)
2. Allow option to chose HTTP client
  a. Switch to a flexible library like Tesla
  b. Add an abstraction in honeybadger (this is how I handle it in https://github.com/pow-auth/assent/tree/main/lib/assent to support built-in httpc by default)

A third option, which is about solving the IPv6 issue, is to switch to something other than hackney since upstream fix seems unlikely to happen anytime soon. I solved my issue by switching to Finch and pass in `conn_opts`: https://github.com/danschultzer/honeybadger-elixir/commit/14bdbe21f6ea754c335478ccb82bb6fa312993a7

**Comment by danschultzer** [comment]
Sounds great @sorentwo, thanks for looking at this!

I did make a quick PR in #468 to pass hackney opts which solves part of the issue. But it would be nice to just pass in an adapter to deal with edge cases and have more control over the connection egress.

Just a warning on `:httpc` that the cert check has to be handled explicitly (not built-in 😞). But apart from that and no HTTP/2 support, it works fine, and the benefit is that it helps limit the dependency graph size.

**Comment by fastjames** [comment]
Is there still any interest in implementing this change? As of today hackney has a low-priority CVE with no fix listed (and no obvious fix activity on github). I'd love to swap it out for a different client if possible.

**Comment by fastjames** [comment]
> [@fastjames](https://github.com/fastjames) Yup!

Well, it's good to know that someone else is interested as well. I read over some of the earlier links on this issue, and @danschultzer 's work in assent looked like it might work here too. I took a stab at shamelessly copying what he did, and I think I'm getting somewhere (see https://github.com/fastjames/honeybadger-elixir/tree/fastjames/feature/multi_http_client_support for my work so far).

Important questions still to be answered:

1. Is hackney support still important? I'm working on that now, but if literally no one wants it then 💥 .
2. How important is pool tuning? The original hackney implementation created its own pool with a connection limit of 30. It looks like `finch` supports pooling (which means `req` does), but for `:httpc` we would have to provide our own pool implementation.

**Comment by sorentwo** [comment]
> Is there still any interest in implementing this change?

I'm still interested in this change, but I dropped the ball implementing it.

@fastjames Thanks for taking a shot at this. As to your questions:

> Is hackney support still important? I'm working on that now, but if literally no one wants it then 

It should be possible to use other adapters. A `httpc` base is nice, but people should be able to use `Req` or other upcoming options.

> How important is pool tuning?

Switching to `httpc` without a pool option could cause a problem in clients that experience a high volume of errors. This is tricky to answer, because using a custom pooler would negate most of the benefit of using `httpc` over `finch`.

**Comment by fastjames** [comment]
> It should be possible to use other adapters. A `httpc` base is nice, but people should be able to use `Req` or other upcoming options.
> 
> > How important is pool tuning?
> 
> Switching to `httpc` without a pool option could cause a problem in clients that experience a high volume of errors. This is tricky to answer, because using a custom pooler would negate most of the benefit of using `httpc` over `finch`.

Thanks for the feedback! Based on these answers, it sounds like the best course of action is to do the following:

1. For now, drop `httpc` as an out-of-the-box supported client. Someone can add it later, but the complexity might get in the way of shipping the larger multi-client feature.
2. Keep hackney in for now, at least for historical purposes.
3. Ensure that each supported client has a proper pool configuration.

**Comment by sorentwo** [comment]
> Thanks for the feedback! Based on these answers, it sounds like the best course of action is to do the following:

I agree with your assessment, except for _ensuring_ each supported client has a pool. As long as we ensure the default has a pool for backward compatibility, people should be able to plug in any other http adapter they like. WDYT?

**Comment by fastjames** [comment]
> > Thanks for the feedback! Based on these answers, it sounds like the best course of action is to do the following:
> 
> I agree with your assessment, except for _ensuring_ each supported client has a pool. As long as we ensure the default has a pool for backward compatibility, people should be able to plug in any other http adapter they like. WDYT?

That's fair, and a good narrowing of scope (which I always like).

Since `req` seems to be pretty solid and since it already handles pools for HTTP/1, I would like to propose we make that the default. Does that sound reasonable? The elixir application I'm currently working on is not high-volume so I would like to have input from others who may be more affected by this change.

**Comment by fastjames** [comment]
I have opened the PR. I'm happy to squash it if everyone is OK with dropping `:httpc` for this iteration.
